File: android/app/src/main/kotlin/org/fgsec/halogen/core/JournalMutation.kt
package org.fgsec.halogen.core

import kotlinx.coroutines.launch
import org.fgsec.halogen.wire.EpisodeData
import org.fgsec.halogen.storage.CacheKey
import org.fgsec.halogen.storage.OutboxOp
import org.fgsec.halogen.components.ToastCenter

/** Persist intent before changing cached state; a profile switch cancels the UI update. */
suspend fun HalogenCore.ensureQueued(operation: OutboxOp.Kind): Boolean {
    val queue = outbox
    if (queue == null) {
        ToastCenter.error("Sync storage is unavailable. This change was not saved.")
        return false
    }
    return queue.enqueue(operation) && outbox === queue
}

fun HalogenCore.enqueueMutation(operation: OutboxOp.Kind, episode: EpisodeData? = null, apply: () -> Unit) {
    val queue = outbox
    val snapshotStore = store
    scope.launch {
        if (queue == null) {
            ToastCenter.error("Sync storage is unavailable. This change was not saved.")
        } else {
            try {
                if (episode != null) snapshotStore?.saveDurably(episode, CacheKey.episode(episode.id))
                if (queue.enqueue(operation) && outbox === queue) apply()
            } catch (error: kotlinx.coroutines.CancellationException) { throw error }
            catch (_: Exception) { ToastCenter.error("Couldn't save the episode. Please try again.") }
        }
    }
}

/** Compound actions persist all operations in the same journal transaction. */
suspend fun HalogenCore.ensureQueuedBatch(operations: List<OutboxOp.Kind>): Boolean {
    val queue = outbox
    if (queue == null) {
        ToastCenter.error("Sync storage is unavailable. This change was not saved.")
        return false
    }
    return queue.enqueueBatch(operations) && outbox === queue
}

File: ios/Halogen/Core/JournalMutation.swift
import Foundation

extension HalogenCore {
    func forAccount(_ originStore: LocalStore?) throws -> HalogenCore {
        guard store === originStore else { throw CancellationError() }
        return self
    }

    /// Persist intent before deriving UI state, and ignore completions from a prior account.
    func ensureQueued(originStore: LocalStore? = nil, _ kind: OutboxOp.Kind) async -> Bool {
        await ensureQueuedBatch(originStore: originStore, [kind])
    }

    func ensureQueuedBatch(originStore: LocalStore? = nil, _ kinds: [OutboxOp.Kind]) async -> Bool {
        if let originStore, store !== originStore { return false }
        guard let queue = outbox else {
            ToastCenter.shared.error("Sync storage is unavailable. This change wasn't saved.")
            return false
        }
        let saved = await queue.enqueueBatch(kinds)
        return saved && outbox === queue
    }

    func enqueueMutation(
        originStore: LocalStore? = nil, _ kind: OutboxOp.Kind, episode: EpisodeData? = nil,
        update: @escaping @MainActor () -> Void
    ) {
        if let originStore, store !== originStore { return }
        let origin = outbox
        let cache = store
        Task { @MainActor in
            guard let origin, outbox === origin else { return }
            if let episode, let cache {
                do { try await cache.saveDurably(episode, key: CacheKey.episode(episode.id)) } catch {
                    ToastCenter.shared.error("Couldn't save episode data. Please try again.")
                    return
                }
            }
            guard outbox === origin, await ensureQueued(kind) else { return }
            update()
        }
    }
}
